services/checks/checks.go

```diff
@@ -131,13 +144,302 @@ func (s *Service) Run(ctx context.Context) {
+	s.cm.Lock()
+	defer s.cm.Unlock()
 
 	r := make([]check.Check, 0, len(s.checks))
 	return append(r, s.checks...)
 }
 
+func (s *Service) checkResults(ctx context.Context) {
+	ticker := time.NewTicker(time.Minute)
+
+	for {
+		select {
+		case <-ticker.C:
+			// continue with next loop iteration
+		case <-ctx.Done():
+			return
+		}
+
+		for id := range s.getResults() {
+			res, err := models.FindActionResultByID(s.db.Querier, id)
+			if err != nil {
+				s.l.Errorf("Can't find acion result: %s.", err)
+			}
+
+			if !res.Done {
+				continue
+			}
+
+			if res.Error != "" {
+				s.l.Errorf("Action %s failed: %s.", id, res.Error)
+				s.removeResult(id)
+				continue
+			}
+
+			_, err = agentpb.UnmarshalActionQueryResult([]byte(res.Output))
+			if err != nil {
+				s.l.Errorf("Failed to parse action result with id: %s, reason: %s.", id, err)
+				s.removeResult(id)
+				continue
+			}
+
+			// TODO Execute script against returned data
+			// TODO Throw away results with expired TTL (they probably should have TTL)
+			//fmt.Println(rr)
+			s.removeResult(id)
+		}
+	}
+}
+
+func (s *Service) addResults(res []checkTask) {
+	s.tm.Lock()
+	defer s.tm.Unlock()
+
+	for _, r := range res {
+		s.tasks[r.resultID] = r
+	}
+}
+
+func (s *Service) removeResult(id string) {
+	s.tm.Lock()
+	defer s.tm.Unlock()
+
+	delete(s.tasks, id)
+}
+
+func (s *Service) getResults() map[string]checkTask {
+	s.tm.Lock()
+	defer s.tm.Unlock()
+
+	res := make(map[string]checkTask, len(s.tasks))
+	for k, v := range s.tasks {
+		res[k] = v
+	}
+
+	return res
+}
+
+func (s *Service) executeChecks() {
+	mySQLChecks, postgreSQLChecks, mongoDBChecks := s.groupChecksByDB(s.checks)
+
+	mySQLRes, err := s.executeMySQLChecks(mySQLChecks)
+	if err != nil {
+		s.l.Errorf("Failed to execute mySQL checks: %s.", err)
+	}
+	s.addResults(mySQLRes)
+
+	postgreSQLRes, err := s.executePostgreSQLChecks(postgreSQLChecks)
+	if err != nil {
+		s.l.Errorf("Failed to execute postgreSQL checks: %s.", err)
+	}
+	s.addResults(postgreSQLRes)
+
+	mongoDBRes, err := s.executeMongoChecks(mongoDBChecks)
+	if err != nil {
+		s.l.Errorf("Failed to execute mongoDB checks: %s.", err)
+	}
+	s.addResults(mongoDBRes)
+}
+
+func (s *Service) executeMySQLChecks(checks []check.Check) ([]checkTask, error) {
+	var res []checkTask
+
+	agents, services, err := s.findAgentsAndServices(models.MySQLdExporterType)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to find proper agents and services")
+	}
+
+	for _, agent := range agents {
+		pmmAgentID := *agent.PMMAgentID
+		r, err := models.CreateActionResult(s.db.Querier, pmmAgentID)
+		if err != nil {
+			s.l.Errorf("Failed to prepare action result for agent %s: %s.", pmmAgentID, err)
+			continue
+		}
+		dsn := agent.DSN(services[*agent.ServiceID], 2*time.Second, "")
+
+		for _, c := range checks {
+			switch c.Type {
+			case check.MySQLShow:
+				if err := s.registry.StartMySQLQueryShowAction(context.Background(), r.ID, pmmAgentID, dsn, c.Query); err != nil {
```

Comment:
And then wait for result synchronously, in the same goroutine. There no need to go async yet.